xdsl/dialects/irdl/irdl.py

```diff
@@ -35,12 +38,47 @@
     SymbolTable,
 )
 from xdsl.utils.exceptions import VerifyException
+from xdsl.utils.str_enum import StrEnum
 
 ################################################################################
 # Dialect, Operation, and Attribute definitions                                #
 ################################################################################
 
 
+class VariadicityEnum(StrEnum):
+    Single = "single"
+    Optional = "optional"
+    Variadic = "variadic"
+
+
+@irdl_attr_definition
+class VariadicityAttr(EnumAttribute[VariadicityEnum], SpacedOpaqueSyntaxAttribute):
+    name = "irdl.variadicity"
+
+
+@irdl_attr_definition
+class VariadicityArrayAttr(ParametrizedAttribute, SpacedOpaqueSyntaxAttribute):
+    name = "irdl.variadicity_array"
+
+    value: ParameterDef[ArrayAttr[VariadicityAttr]]
+
+    def __init__(self, variadicities: Iterable[VariadicityEnum]) -> None:
+        array_attr = ArrayAttr(map(lambda x: VariadicityAttr(x), variadicities))
```

Comment:
Please convert the maps to tuples, they're noticeably faster and I would argue easier to read
```suggestion
        array_attr = ArrayAttr(tuple(VariadicityAttr(x) for x in variadicities))
```